src/Raven.Server/Documents/Handlers/Processors/BulkInsert/AbstractBulkInsertHandlerProcessor.cs

```diff
@@ -0,0 +1,281 @@
+
+    protected async ValueTask<MergedBatchCommand.AttachmentStream> WriteAttachmentStream(long size, Stream stream)
+    {
+        var attachmentStream = new MergedBatchCommand.AttachmentStream();
+
+        if (size <= 32 * 1024)
+        {
+            attachmentStream.Stream = new MemoryStream();
+        }
+        else
+        {
+            AttachmentStreamsTempFiles ??= new List<StreamsTempFile>();
+
+            var attachmentStreamsTempFile = GetTempFile();
+
+            attachmentStream.Stream = attachmentStreamsTempFile.StartNewStream();
+
+            AttachmentStreamsTempFiles.Add(attachmentStreamsTempFile);
+        }
+
+        attachmentStream.Hash = await CopyAttachmentStream(stream, attachmentStream.Stream);
+        
+        await attachmentStream.Stream.FlushAsync();
+
+        return attachmentStream;
+    }
+
+    protected abstract StreamsTempFile GetTempFile();
+
+    protected abstract ValueTask<string> CopyAttachmentStream(Stream sourceStream, Stream attachmentStream);
+
+    protected abstract (long, int) GetSizeAndOperationsCount(TCommandData commandData);
+
+    public BulkOperationResult OperationResult { get; private set; }
+
+    public override async ValueTask ExecuteAsync()
+    {
+        try
+        {
+            IDisposable currentCtxReset = null, previousCtxReset = null;
+
+            try
+            {
+                using (ContextPool.AllocateOperationContext(out JsonOperationContext context))
+                using (context.GetMemoryBuffer(out var buffer))
+                {
+                    currentCtxReset = ContextPool.AllocateOperationContext(out JsonOperationContext docsCtx);
+                    var requestBodyStream = RequestHandler.RequestBodyStream();
+
+                    using (var reader = GetCommandsReader(context, requestBodyStream, buffer, _cts.Token))
+                    {
+                        await reader.Init();
+
+                        var array = new TCommandData[8];
+                        var numberOfCommands = 0;
+                        long totalSize = 0;
+                        int operationsCount = 0;
+
+                        while (true)
+                        {
+                            using (var modifier = new BlittableMetadataModifier(docsCtx))
+                            {
+                                var task = reader.GetCommandAsync(docsCtx, modifier);
+                                if (task == null)
+                                    break;
+
+                                _cts.Token.ThrowIfCancellationRequested();
+
+                                // if we are going to wait on the network, flush immediately
+                                if ((task.Wait(5) == false && numberOfCommands > 0) ||
+                                    // but don't batch too much anyway
+                                    totalSize > 16 * Voron.Global.Constants.Size.Megabyte || operationsCount >= 8192)
+                                {
```

Comment:
I think those conditions are still applicable in sharded env. 

Its about sending the buffered commands while we're waiting for the network to read - first part: `(task.Wait(5) == false && numberOfCommands > 0)`. And the second part is about sending the commands in batches instead of single command one by one. I think that then the shard can read and process them more efficient.

Reply:
After talking to Karmel we decided that we won't change it until we do the bulk insert testing and have some performance numbers. 

https://issues.hibernatingrhinos.com/issue/RavenDB-18453